platform-sdk/docs/proposals/state-refactor/state-refactor.md

```diff
@@ -0,0 +1,211 @@
+* `PlatformBuilder` already have a mechanism to delegate genesis state creation to the app. A similar mechanism can be provided to delegate loading existing state from a stream to the app.
+* `Hedera` will implement this mechanism by creating a `MerkleHederaState` object from a protobuf stream - see "Reading from protobuf" section above - and providing it access to all registered codecs using the services registry
+* When `MerkleHederaState` reads its state children from the stream, it will first read state metadata (service name and state key from the label, state type may be hard-coded?), then find the codecs, and provide them to the state children
+* For example: `TokenService.ACCOUNTS` state is loaded. Service name is `TokenService`, state key is `ACCOUNTS`, state type is “map”. Key and value codecs for this service/state are loaded. Using these codecs, a MerkleDb data source builder is created, then created a virtual map from the input stream and the builder
+
+## Merkle States Root
+
+There are two proposed changes to the current tree structure.
+
+The **first change:**
+
+![new-merkle-tree-structure.png](assets/new-merkle-tree-structure.png)
+
+Current `MerkleHederaState` must become a root, `PlatformState` will be moved to become its child, and then `State` will be removed.
+
+The **second change** to the root that is required for optimal state proofs. It is necessary to turn the top level n-ary tree (root + named states) into a virtual binary tree inside containing all the named states as leaves. Ideally it will store those virtual internal nodes all in a single array of hashes. When serialized it can save out the array of hash and the protobuf messages for each of its children.
+
+## Platform State as a Singleton
+
+### Singleton State
+
+A singleton is a node with 2 children:
+
+* String label for State Name
+* Protobuf message for child data
+
+When writing to disk it should have a SingletonState protobuf message with 3 fields:
+
+* Hash of children Hashes
+* String label for State Name
+* Protobuf message for child data. This field is problematic as it can be one of many kinds. It seems the best solution is probably using protobuf `any` type. The other option is `oneof` with all the singleton content types. See:
+  * [https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/any.proto](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/any.proto) ## Saved State Snapshot Changes
+
+Currently [PlatformState](/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java) is a special case node of the merkle tree. As a part of the simplification effort **it should be refacored to a set of singleton states**. Also, some of the `PlatformState code` is no longer relevant, and needs to be removed. There is a prerequsite for the `PlatformState` to become a singleton - Platform State split
```

Comment:
You have the following in code ticks:`PlatformState code`

but I assume you intended to just have `PlatformState` in ticks

Reply:
Agreed. I removed the link